### ISSUE TITLE
Code cleanup

### ISSUE THREAD

**Issue by lweberk** [Return-To-The-Roots/s25client#94]
- [ ] Reduction / Removal of macros by pushing platform dependant stuff onto interface / implementation modules chosen by the buildsystem / meta-buildsystem.
- [ ] Compaction of code by central instance of typedefs for shorts like uint, ushort, u<anything "unsigned">
- [ ] Horizontal alignment whenever it eases readability.
- [ ] Reduction of singletons whenever possible.  

+ Anything that you can come up with... please share and discuss. This can be a controversial topic and needs proper discussion.

**Comment by Flow86** [comment]
about point 2: Compaction of code by central ...
theres a "stdint(.h)" for that - no need to typedef ourselfs

**Comment by MarcusSt** [comment]
I'd call it an enhancement, since the nightly has been running almost flawlessly up to (and including) 2015-07-27. The game would run for hours without crashes or becoming async. We should _first_ restore that state, since there are people actually playing the game and the latest stable isn't exactly that stable/async-free.

**Comment by Flow86** [comment]
thats definitly the priority nr 1

**Comment by lweberk** [comment]
@Flow86 
You are right. Did not realize it was taken into the standard at C++11. Should we also expand to the usage of explicit sizes? Like int16_t for signed short instead of "short"?

@MarcusSt 
Yes it definitively is. The goal is to take a different approach to the whole "rewrite" initiative by not breaking things while partially doing so (which otherwise would never take of). A challenge to refurbish the house without it crumbling, so to speak.
The absolute requirement is just that, for every cleanup, make sure nothing broke.

**Comment by Flow86** [comment]
@Iweberk
Its not C++11, its already in older C++.
But exactly, thats what I think of: int => (u)int32_t, short => (u)int16_t etc.
We only have to be careful not to break things ;-)

**Comment by Flamefire** [comment]
I think this can be done by a global S&R (the type conversion). And we should set up some pre-merge tests. So for example an auto-formatter tool, which enforces formatting stuff (newlines after ";", spaces after "," [sometimes there is some very ugly code...], spaces instead of tabs, etc) What I have seen in other projects: a tool, that checks formatting on commit/pre-merge and refuses it, if there is a diff between the commit and the commit run through the formatter. So the submitter has to rum the formatter first.

**Comment by lweberk** [comment]
@Flow86 
Yepp, got the gist of it. Sidenote: AFAIK it entered the standard just on C++11, it was *supported* but not standardized before that. Could be wrong about it though, but overall that was my primary concern.

@Flamefire 
You mean something like a linter? True. Useful tools, but in my opinion cumbersome when you start "painting" the code (horizontal indentation being an example of that).
As for spaces vs tabs. Spaces are a feature not a bug. If you indent horizontally you run into hell if you use tabs. Thus I'd keep that current choice of style the way it is (4 hard spaces)

**Comment by Flamefire** [comment]
Yes. But there are powerful tools that will just do that. it also reduces merge conflicts (one did use spaces, the other tabs e.g.)
And this is really something the should be done by a tool, not by hand. E.g. checking IF it has been done is easy then.
What do you mean with your last paragraph? Tabs should NEVER be used in the code... So always use 4 spaces for indendation

**Comment by lweberk** [comment]
@Flamefire 
I think we are all consenting adults. In the end its about people being responsible for configuring their working environments properly for the project. The experience I've had with automating enforcement, is that people grow complacent and start to rely on these kinds of automations. It leads to inherent messiness in the way they do things. A little like the reliance on IDE's, they tend to nurture bad design... because you can get "away with it". If one has no such tool as an IDE, one has to be more careful about how and what they where doing, so not to get lost in ones own mess.

Agree on spaces, seems I misunderstood your phrasing above.

**Comment by Flamefire** [comment]
But especially for large commits, you don't want to check every space that gets added. E.g. in one of Maqs commits a tab-issue slipped through. Tools are designed to fulfil tasks that are repetitive or very stressful for humans. So why not use auto-formatter tools? Why not rely on them? If they actually did something wrong, mergers can still catch that (in case someone "relied to much on them")
And about IDEs: I understand your opinion. We have this discussion every now and then. So why don't you write your code with vi/notepad only? Or better: Burn them on disk with magnets? Don't rely on any tool! ;-) I think: if there are tools to simplify something: Use it! If they have shortcomings, find them, fix them, or note them. If they don't work, don't use them.
I don't see anything wrong in people writing "poorly" formatted code. They should be able and allowed to use whatever style they wish. Only in commits the style should be fixed. And tools do a great deal for that: If I want my own style, I check a commit out, run it through my formatter, change it, run it through the projects formatter and commit. Everyone is happy ;)

**Comment by MarcusSt** [comment]
The thing is that I'm working on a lot of different projects. Some prefer spaces, others prefer tabs and it just takes ages to change the settings every single time. I'm not using an IDE. I just use a simple editor with syntax highlighting. Therefore I'd also go for an automated solution. At least replacing tabs with spaces at the beginning of a line should be no biggie.

**Comment by Flamefire** [comment]
Add some other things like 1 space after commas and newlines after ";" Also simple to do

**Comment by lweberk** [comment]
There is no disagreeing with the usage of auto formatters, but they *remain* the responsibility of the one pushing the change, not the one merging it in. There should not be the burden to do so for any repository steward. To change the phrasing; each end everyone should wipe his own ass, and not ask or imply others to do it for them. How someone chooses to do so, remains entirely up to the individual. Again, we are all consenting adults.
As for the IDE's. Its about cross-project auto-completion. As far as a decent editor goes, actually I agree to use something that solves your problem.

**Comment by Flamefire** [comment]
This is right. What I meant is having a commit hook, that checks, IF the stuff is correctly formatted to the extent, this can be automaticly checked. This gives the merger the chance, to deny a non-formatted commit without checking all the changes. It should not replace additional checking. Just extent and simplify it.

**Comment by lweberk** [comment]
Agreed, it still forces the commiter to fix their stuff without getting accustomed of sloppiness getting trough.

**Comment by lweberk** [comment]
Should I touch char/unsigned char/signed char?? They are not standardized... whats worse it seems to be pretty arbitrary. Run into this, because using int8_t and uint8_t is starting to uncover implicit castings.

**Comment by Flamefire** [comment]
If they are not standardized, I'd say remove them. Could cause asyncs

**Comment by Flamefire** [comment]
Add this to the list: Remove as many pointers as possible. Use boost::array, vector, references, boost:scoped_ptr or at very last boost::intrusive_ptr or even (worse) boost::shared_ptr

**Comment by lweberk** [comment]
@Flamefire done.

**Comment by Flamefire** [comment]
Can you push your code to your repo so we can see what you are doing? :)

**Comment by lweberk** [comment]
Sure. Non-compiling at current state though.

**Comment by Flamefire** [comment]
Then general advice: Try to make small commits that compile. Don't make to many changes in one commit as it makes it hard to review and recheck (go back, test earlier versions). For some it is not possible but e.g. replacing integer types could be one commit that compiles standalone.
Also wait for my next Terrainrenderer commit before you do that one as it almost completely changes the code already using better data structures.
And one advice I got from maqs: For non-logic changing commits, create a replay that uses a lot of features and runs some time. Run that one after the changes and if it goes async you missed something. Good and easy way to check a lot of stuff.

**Comment by lweberk** [comment]
https://github.com/lweberk/s25client/tree/fixed_width_types

Its a throw away branch, since it touches too much to be able to keep up with current master. One of the rationales of #145. Same would go with astyle; works really nice, but touches ~50k of lines.

The branch itself is generated by; https://gist.github.com/lweberk/8d9eada92ef725585b51 . Sorry for the repeating command, was faster to focus on the regexes and duplicate lines in the editor than making the whole thing sophisticated.

**Comment by Flamefire** [comment]
Alright. Can we maybe get a vote from all devs if anyone is doing something big or if we can run astyle now?
And does anyone know how to set up an automatic check if it was run before commit/merge?

**Comment by MarcusSt** [comment]
1. I'd prefer not to lose the blame functionality for now, which will most likely happen with astyle. Once we are sure there are no bigger issues remaining, this could be done.
2. No idea, sorry.

**Comment by Flow86** [comment]
lets wait for both (uintX_t-stuff and astyle) until end of august, I think we could fix the remaining problematic issues until then.

**Comment by lweberk** [comment]
@Flamefire 
git hooks, you can hook up scripts to do that. 

@MarcusSt @Flow86 
Will focus on translation then. Anybody against it? ...still need to familiarize with the codebase, which destroys my return of investment for my time if I just go jumping all over the place hunting a bug...

**Comment by Flow86** [comment]
yes git hooks work, but are run on client side (I think github dont has support for server side hooks)

Translation of the source code comments, or which translations do you mean?

**Comment by lweberk** [comment]
Yes. Translation of comments.

**Comment by Flamefire** [comment]
Something that should also be done: I introduced a class enum "Direction" which is used in the AI only ATM. This ensures correct usage of direction values and makes the code easier to read. There are a lot of things like "GetNeighbour(pt, 3)" <-- WTF is 3? (Would someone think that reads the code)
It should be enforced to use the direction: GetNeighbour(pt, Direction::SOUTH_WEST(?)) for example. Also all iterations would then be better by using the COUNT member. I'd also add an iterator member to the Direction class, so that you can iterate from ANY direction one time through all of them. Would need some thinking though.

**Comment by lweberk** [comment]
Added the point. I think the best way for that one is to just "boyscout" it.

**Comment by Flamefire** [comment]
What you mean by boyscout? I'd change the interfaces (GetNeighbour*()) and fix the code till it compiles ;) Can be a S&R mostly but got to invent an iterator system.

**Comment by lweberk** [comment]
Boyscouting means; leave the place behind in a cleaner state as it was before. It does not imply targeted cleaning up.

**Comment by Flamefire** [comment]
I added naming scheme to the list. Can be done by somebody without much code knowledge as its a simple refactoring (use an IDE for that!)
What I mean is use camelCase/CamelCase for all names. Remove the underscore-versions and the (very ugly) lowercase versions like tmposroopsOnMission or even worse ones...

**Comment by lweberk** [comment]
Additional Note: usage of "this->" for the sake of scope explicitness?

**Comment by Flamefire** [comment]
This would clutter the code to much and is against the lazy programmers habit ;)
But I'd add namespaces for all new stuff. With namespace = folder structure (and maybe even rttr as global namespace)

**Comment by lweberk** [comment]
:D. Hopefully it won't hurt people if I accidentally decide to add them...
Namespaces make sense, but would not use the top-level one, since it will not be included in another application, like a library would.

**Comment by Flamefire** [comment]
Oh right. But the libraries should use them

**Comment by Flow86** [comment]
yea but not a rttr namespace, since the libraries dont have anything to do with rttr (besides of liblobby)

**Comment by lweberk** [comment]
They already do, if I'm not mistaken? At least as of consulting with grep.

...which showed me something else to add to the list; remove arbitrary namespace imports (using namespace)

**Comment by Flow86** [comment]
yes most of them should already have a namespace. and yes, using namespace  ist most of the time ugly

**Comment by Flamefire** [comment]
Changed the milestone as this is to much work till the next release. Any progress towards this is appreciated but I'd concentrate on current bugs and mem-leaks etc. till the 0.8.3
